app/models/user.rb

```diff
@@ -84,9 +84,8 @@ def moderatable
   validates :email, format: { :with => RE_EMAIL_OK, :allow_blank => false,
                               :message => MSG_EMAIL_BAD, :unless => :minimal_signup? }
 
-  #TODO: this needs tests
```

Comment:
I'm not sure we can remove this TODO, because this PR doesn't add tests, so if it needed tests before, it still needs them.

I think we would need some tests to check whether the correct validation process is triggered, or in other words, we need tests for the `if` statements in the weak and strong validations:

weak:
` if: -> { validate_password? && !provider_requires_strong_passwords? }`

strong:
`:if => :provider_requires_strong_passwords?`

We got these three situations:

1. `validate_password?` is `false`, no validation triggered
2. `validate_password?` is `true`, weak validation required
3. `validate_password?` is `true`, strong validation required

However, `user_test.rb` is so long and contains so many tests I can't tell if such logic is actually tested. I'd say it's not. There are some tests for weak and strong passwords. Those tests assume the correct validation triggers and verify it works fine, but I couldn't find tests to verify the validation triggering logic itself. Which as I understand it, is what that TODO is about.

On the other hand, I think we can remove the `:allow_blank => false`, because the whole validation will be skipped if `password_required?` and therefore `validate_password?` are `false`. So, it seems we are explicitly allowing blank passwords when `validate_password?` is `true`, is that intentional? Why setting a min size of 6 then?

Reply:
Yeah, you're right...
Although https://github.com/3scale/porta/pull/3646/files#diff-987571c214a9915aebd953182b9437217a7f4dc85f57e49b86fb004c6e66cfd4R83 adds a test that verifies specifically this validation (when the conditions are matched), as well as https://github.com/3scale/porta/pull/3646/files#diff-04d6f0b762fc398af5bc11ebc17d8b2ca034a95637f7ea0eb4847f6588ab67cbR778
It is true though that it's not clear, and probably some of the branches you mention might indeed be missed, I'll double-check.

As for `allow_blank`, I think it's actually wrong, and shouldn't be there... But I was not sure about it, and was afraid to break something, so I left it as it was :shrug: Probably that's for some future PRs :sweat_smile:

Reply:
I thought it was something for non-password login..

Reply:
I added some more explicit tests here:
https://github.com/3scale/porta/pull/3646/commits/b6a7c24baa0b9abf4a863a4750a4eebffd555057

Reply:
Also, I moved the two `validates :password` lines closer together, so it's more clear what happens (having them separate was confusing)